### DISCUSSION TITLE
Primefaces 14 - @widgetVar behavior

### DISCUSSION THREAD

**Discussion by Bimnomercy** [primefaces/community#2825] [PrimeFaces] [answered]
Hello, I am migrating a project from version 13 to version 14.
I am having a lot of problems with the ajax update using the @widgetVar tag. Most of them cause a ComponentNotFoundException when rendering, while these pages worked fine in version 13

I am looking for documentation on the behavior changes of the @widgetVar tag in version 14.

Can someone tell me how to do this?

Thanks in advance.

**Comment by Bimnomercy** [comment]
I'm starting to understand after a series of tests. The operation of update="@widgetVar(stuff)" is a real mess in version 14.

For example, the following code:

```
<p:tabView dynamic="true">
	<p:tab title="Header I">

	</p:tab>

	<p:tab title="Header II">
		<p:commandButton value="Add 1" action="#{testView.add()}" update="@widgetVar(cptLabel)" />

		<p:fieldset widgetVar="cptLabel">
			<p:outputLabel value="#{testView.cpt}" />
		</p:fieldset>

	</p:tab>
</p:tabView>
```

...causes a ComponentNotFoundException because obviously the "cptLabel" fieldset is not yet rendered when the commandButton is.
**This problem did not occur in V13.**

Another example, the code:

```
<h:form>
	<p:dialog widgetVar="dialogModification" modal="true" resizable="false" closeOnEscape="false">

		<p:outputLabel value="#{view.selection.libelle}" />
		
	</p:dialog>
</h:form>
.
.
.
<h:form>
	<p:dataTable widgetVar="datatable" value="#{view.tableau}" var="ligne" size="small">

		<p:column styleClass="center" width="5px;">
			<div style="width:5px;height:30px;background-color: #{ligne.couleur};" />
		</p:column>

		<p:column styleClass="center" width="100px;" sortBy="#{ligne.code}" filterBy="#{ligne.code}" headerText="Code">
			<p:commandLink action="#{view.openDialog(ligne)}" update="@widgetVar(dialogModification)">
				<h:outputText value="#{ligne.code}" styleClass="hyperLink dataTableRow" />
			</p:commandLink>
		</p:column>

		<p:column sortBy="#{ligne.libelle}" filterBy="#{ligne.libelle}" headerText="Libelle">
			<p:commandLink action="#{view.openDialog(ligne)}" update="@widgetVar(dialogModification)">
				<h:outputText value="#{ligne.libelle}" styleClass="hyperLink dataTableRow" />
			</p:commandLink>
		</p:column>

	</p:dataTable>
</h:form>
```

In this case, the lines of the datatable remain white. Remove the "update="@widgetVar(dialogModification)"" or put the "p:dialog" in the same <h:form> as the datatable and everything is back to normal.
**This problem did not occur in V13.**

The operation of update="@widgetVar()" in V14 is a real regression compared to V13. The widgetVar loses all the flexibility of use that made it essential in Primefaces.

I read that the @widgetVar engine has been replaced by the native jsf search expressions in V14. I sincerely think that this is a mistake and that we need to go back.

In any case, I am not ready to migrate my projects to V14.

**Reply by tandraschko** [reply]
create a issue pls and attach a forked primefaces-test reproducer, otherwise noone can help.

**Comment by Bimnomercy** [comment, marked as answer]
I am making progress in my investigations. In fact, when an update="@widgetVar(stuff)" attribute is rendered in the JSF page while the component with the widgetVar="stuff" attribute is not yet, we can have a strange behavior of the page or a ComponentNotFoundException.

The attached project shows the problem, the labels of the datatable of tab #1 are not displayed.

The test page is a "ui:composition". You will find in the "template.xhtml" page some commented code that allows for normal operation. The problem only occurs when the datatable is in a tabview. Removing the "update="@widgetVar(dialogDetail)" on line 46 of the test.xhtml file no longer causes the bug either, strange...

[TestPrimefaces.zip](https://github.com/user-attachments/files/17309227/TestPrimefaces.zip)

**Comment by tandraschko** [comment]
i had a short look but no idea. as i said, please create a issue and attach the example.
the weird stuff is that its actually the same implementation of the `@widgetVar` expression, just the surrounding classes changed.
i think it requires more work to debug, maybe you need to invest time yourself or sponsor a fix.

**Reply by melloware** [reply]
Ticket opened:https://github.com/primefaces/primefaces/issues/12701
